primitives/session/src/lib.rs

```diff
@@ -106,6 +108,69 @@ impl GetValidatorCount for MembershipProof {
 	}
 }
 
+/// A trait for online node inspection in a session.
+///
+/// This trait is used to decouple the pallet-session dependency from im-online
+/// module so that the user of im-online & offences modules can pass any list of
+/// validators that are considered to be online in each session, particularly useful
+/// for the Substrate-based projects having their own staking implementation
+/// instead of using pallet-staking directly.
+pub trait ValidatorSet<AccountId> {
+	// TODO [ToDr] This could use `frame_support::Parameter` instead,although don't know if such
+	// import is legal.
+	type ValidatorId: codec::Codec + codec::EncodeLike + Clone + Eq + sp_std::fmt::Debug;
+	// TODO [ToDr] This is most likely not needed along with `AccountId`
+	type ValidatorIdOf: Convert<AccountId, Option<Self::ValidatorId>>;
+
+	/// Returns current session index.
+	fn current_index() -> SessionIndex;
```

Comment:
Can we please rename this to `seession_index`, reading the code in some other place it looks like it's `ValidatorSet` index (`let session_index = T::ValidatorSet::current_index()` looks weird).